crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -1,260 +1,524 @@
+            )
+            .add_system_set(
+                State::on_exit_set(MyState::S2).with_system(
+                    (|mut r: ResMut<Vec<&'static str>>| r.push(dbg!("exit S2"))).system(),
+                ),
+            )
+            .add_system_set(State::on_enter_set(MyState::S3).with_system(
+                (|mut r: ResMut<Vec<&'static str>>| r.push(dbg!("enter S3"))).system(),
+            ))
+            .add_system_set(
+                State::on_update_set(MyState::S3).with_system(
+                    (|mut r: ResMut<Vec<&'static str>>, mut s: ResMut<State<MyState>>| {
+                        r.push(dbg!("update S3"));
+                        s.overwrite_push(MyState::S4).unwrap();
+                    })
+                    .system(),
+                ),
+            )
+            .add_system_set(State::on_pause_set(MyState::S3).with_system(
+                (|mut r: ResMut<Vec<&'static str>>| r.push(dbg!("pause S3"))).system(),
+            ))
+            .add_system_set(
+                State::on_update_set(MyState::S4).with_system(
+                    (|mut r: ResMut<Vec<&'static str>>, mut s: ResMut<State<MyState>>| {
+                        r.push(dbg!("update S4"));
+                        s.overwrite_push(MyState::S5).unwrap();
+                    })
+                    .system(),
+                ),
+            )
+            .add_system_set(State::on_inactive_update_set(MyState::S4).with_system(
+                (|mut r: ResMut<Vec<&'static str>>| r.push(dbg!("inactive S4"))).system(),
+            ))
+            .add_system_set(
+                State::on_update_set(MyState::S5).with_system(
+                    (|mut r: ResMut<Vec<&'static str>>, mut s: ResMut<State<MyState>>| {
+                        r.push(dbg!("update S5"));
+                        s.overwrite_push(MyState::S6).unwrap();
+                    })
+                    .system(),
+                ),
+            )
+            .add_system_set(State::on_inactive_update_set(MyState::S5).with_system(
+                (|mut r: ResMut<Vec<&'static str>>| r.push(dbg!("inactive S5"))).system(),
+            ))
+            .add_system_set(
+                State::on_update_set(MyState::S6).with_system(
+                    (|mut r: ResMut<Vec<&'static str>>, mut s: ResMut<State<MyState>>| {
+                        r.push(dbg!("update S6"));
+                        s.overwrite_push(MyState::Final).unwrap();
+                    })
+                    .system(),
+                ),
+            )
+            .add_system_set(State::on_resume_set(MyState::S4).with_system(
+                (|mut r: ResMut<Vec<&'static str>>| r.push(dbg!("resume S4"))).system(),
+            ))
+            .add_system_set(State::on_exit_set(MyState::S5).with_system(
+                (|mut r: ResMut<Vec<&'static str>>| r.push(dbg!("exit S4"))).system(),
+            ));
+
+        const EXPECTED: &[&[&str]] = &[
+            &["startup", "update S1"],
```

Comment:
It looks like we have lost same-frame state transitions, which I would likne to preserve if possible. Can you explain the motivation and/or limitations encountered here (sorry if this has been discussed elsewhere or if i forgot a previous conversation ... lots going on right now :smile: )